### ISSUE TITLE
Training steps and training time have doubled

### ISSUE THREAD

**Issue by GunnarQuist** [d8ahazard/sd_dreambooth_extension#699]
### Kindly read the entire form below and fill it out with the requested information.

**Please find the following lines in the console and paste them below. _If you do not provide this information, your
issue will be automatically closed._**

`
Python revision: 3.10.7 (main, Nov 24 2022, 19:45:47) [GCC 12.2.0]
Dreambooth revision: 4ca69a904f5ddd5651d87032b3dca515eea505ba
SD-WebUI revision: cca6e5212ddd77be7efdce2c27e475d3af436d85

Checking Dreambooth requirements...
[+] bitsandbytes version 0.35.0 installed.
[+] diffusers version 0.10.2 installed.
[+] transformers version 4.25.1 installed.
[+] xformers version 0.0.14.dev0 installed.
[+] torch version 1.12.1+cu113 installed.
[+] torchvision version 0.13.1+cu113 installed.
`

**Have you read the [Readme](https://github.com/d8ahazard/sd_dreambooth_extension#readme)?**
Yes.

**Have you completely restarted the stable-diffusion-webUI, not just reloaded the UI?**
Yes.

**Have you updated Dreambooth to the latest revision?**
Yes.

**Have you updated the Stable-Diffusion-WebUI to the latest version?**
Yes.

**No, really. Please save us both some trouble and update the SD-WebUI and Extension and restart before posting this.
Reply 'OK' Below to acknowledge that you did this.**
Yes.

**Describe the bug**

(A clear and concise description of what the bug is)
The Dreambooth extension seems to train for twice as many steps as expected. If I have x number of dataset images and y number of epochs, I'm expecting the number of steps in the training to be (x*y). Instead I'm seeing in the CLI a number equal to (x*y*2).

**Provide logs**

If a crash has occurred, please provide the *entire* stack trace from the log, including the last few log messages *
before* the crash occurred.
There was no crash. The Dreambooth extension trained to completion and produced a checkpoint. It just took twice as long as I expected.

**Environment**

What OS?
Linux

If Windows - WSL or native?
n/a

What GPU are you using?
Nvidia RTX A5000

**Screenshots/Config**
If the issue is specific to an error while training, please provide a screenshot of training parameters or the
db_config.json file from /models/dreambooth/MODELNAME/db_config.json

This is not a screenshot, it is a copy-paste of text output to the CLI:

Sched breakpoint is 3500
  ***** Running training *****
  Instance Images: 70
  Class Images: 1400
  Total Examples: 140
  Num batches each epoch = 140
  Num Epochs = 100
  Batch Size Per Device = 1
  Gradient Accumulation steps = 1
  Total train batch size (w. parallel, distributed & accumulation) = 140
  Total optimization steps = 7000
  Total training steps = 14000
  Resuming from checkpoint: False
  First resume epoch: 0
  First resume step: 0
  Lora: False, Adam: False, Prec: fp16
  Gradient Checkpointing: True, Text Enc Steps: -1.0
  EMA: True
  LR: 1.72e-06)

As you can see, the number of Instance Images is 70 and the Num Epochs is 100. I'm expecting 7000 for Total Training Steps but instead it is double that, 14000. I don't know what Total Optimization Steps are but it has the number of steps I'm expecting for "total steps".

All of this might be a misunderstanding on my part. But the way I seem to perceive it, my training steps and training times have doubled.

**Comment by Straafe** [comment]
I've noticed this too. Also (potentially unrelated), it is no longer saving checkpoints along the way like it used to, despite the setting to save ever X epochs.

**Comment by GunnarQuist** [comment]
> 

I think you have to check "Generate a .ckpt file when saving during training" under the Saving tab in order to have it save checkpoints along the way like it used to. I could be wrong. I haven't tried it yet. I've been pretty confident with my training settings lately so I haven't bothered with trying to do that.

**Comment by Straafe** [comment]
You're right - looks like that new setting saves incrementally, I didn't notice that one

**Comment by ekeric13** [comment]
Getting the double steps issue as well.

Also confused on how to save checkpoints incrementally. @Straafe have you figured it out? I don't get what "save model frequency (epochs)" mean... like if it's 1 does it save it every epoch? if its 2 does it save it every 2 epochs?

**Comment by Straafe** [comment]
@ekeric13 Yea, looks like this setting (which defaults to disables) is will enable the incremental checkpoints being saved:
![image](https://user-images.githubusercontent.com/7477619/210912296-a4720b1d-220e-43fc-987e-7ba56ebc3043.png)

The frequency of the incremental saves for me also does not match up with my epoch/step settings, so not sure about that.

**Comment by Straafe** [comment]
I did notice if you do not use or generate any classification images, it does run for the expected amount of steps based on the input settings, so maybe that's a clue.

**Comment by ekeric13** [comment]
"Total Examples" being 2x "Instance Images" when there is regularization is definitely a thing:

https://github.com/d8ahazard/sd_dreambooth_extension/blob/6034263b1449bdf130d74afe7d661a750e38e390/dreambooth/train_dreambooth.py#L578

not sure how that factors in...

The reason for the doubling seems to be here:
https://github.com/d8ahazard/sd_dreambooth_extension/blob/6034263b1449bdf130d74afe7d661a750e38e390/dreambooth/train_dreambooth.py#L502-L506

`train_dataset.num_train_images` = 20
`len(train_dataloader)` = 40

I think both should be 20. This way the only difference between the number of steps is the batchSize.

**Comment by GunnarQuist** [comment]
> I did notice if you do not use or generate any classification images, it does run for the expected amount of steps based on the settings, so maybe that's a clue.

Yes! I think this may be what is happening to me. My first training had to generate an entirely new set of class images. It ran twice the amount of training steps. But I just completed a second model that I trained using the same generated class images and it ran for the expected number of steps.

**Comment by d8ahazard** [comment]
This should be resolved in the latest version.

Training steps in the UI have been updated to reflect the total number of iterations over the training dataset, excluding gradient accumulation steps, but counting batch size. Previously, it only stepped when the optimizer was stepped, which isn't an accurate representation of training progress, IMHO.

As such, this is the explanation for the shift to epochs across the project - because steps just isn't helpful for anybody involved. It's a vague, random number; whereas epochs should be a better measure of what's actually being trained. (Arguably). 

So, do an update, consider the above, you should be able to see what I'm talking about.

**Comment by ekeric13** [comment]
Excited to try it out!
